### PR TITLE
feat: improve query performance for game and achievement metrics updates

### PR DESCRIPTION
**Pull request by luchaos**
Add indexes for player_achievements, player_games, and Achievements tables to improve game metrics queries.

Migrations have been applied to stage environment.

Refactor `getTotalUniquePlayers` to use `player_achievements`.

Verified that the query optimizer actually uses those indexes using the following queries:
```sql
-- players

EXPLAIN
SELECT
  COUNT(DISTINCT pa.user_id)
FROM
  player_achievements pa
  LEFT JOIN Achievements a ON a.ID = pa.achievement_id
  LEFT JOIN UserAccounts u ON u.ID = pa.user_id
WHERE
  a.GameID = 228 -- 1, 228, 337
  AND a.Flags = 3
  AND u.Untracked = 0;

EXPLAIN
SELECT
  COUNT(DISTINCT pa.user_id)
FROM
  player_achievements pa
  LEFT JOIN Achievements a ON a.ID = pa.achievement_id
  LEFT JOIN UserAccounts u ON u.ID = pa.user_id
WHERE
  a.GameID = 228 -- 1, 228, 337
  AND a.Flags = 3
  AND u.Untracked = 0
  AND pa.unlocked_hardcore_at IS NOT NULL;

EXPLAIN
SELECT
  COUNT(DISTINCT aw.User)
FROM
  Awarded aw
  LEFT JOIN Achievements ach ON ach.ID = aw.AchievementID
  LEFT JOIN UserAccounts ua ON ua.User = aw.User
WHERE
  ach.GameID = 228 -- 1, 228, 337
  AND ach.Flags = 3
  AND ua.Untracked = 0;

-- player games

EXPLAIN
SELECT
  COUNT(*)
FROM
  player_games pg
LEFT JOIN UserAccounts u ON pg.user_id = u.ID
WHERE
  pg.game_id = 1 -- 1, 228, 337
  AND pg.game_id IS NOT NULL
  AND u.Untracked = 0;

-- player achievements

EXPLAIN
SELECT
  COUNT(*)
FROM
  player_achievements pa
LEFT JOIN UserAccounts u ON pa.user_id = u.ID
WHERE
  pa.achievement_id = 342 -- 342, 341, 2253, 4874, 347, 2251, 2298, 2299, 2261, 4933 
  AND pa.achievement_id IS NOT NULL
  AND u.Untracked = 0;

-- player achievements hardcore

EXPLAIN
SELECT
  COUNT(*)
FROM
  player_achievements pa
LEFT JOIN UserAccounts u ON pa.user_id = u.ID
WHERE
  pa.achievement_id = 342 -- 342, 341, 2253, 4874, 347, 2251, 2298, 2299, 2261, 4933
  AND pa.achievement_id IS NOT NULL
  AND u.Untracked = 0
  AND pa.unlocked_hardcore_at IS NOT NULL;

-- awarded

EXPLAIN
SELECT
  COUNT(*)
FROM
  Awarded aw
LEFT JOIN UserAccounts u ON aw.User = u.User
WHERE
  aw.AchievementID = 342 -- 342, 341, 2253, 4874, 347, 2251, 2298, 2299, 2261, 4933
  AND aw.AchievementID IS NOT NULL
  AND u.Untracked = 0;
```

### CODE REVIEW

File: database/migrations/platform/2023_10_09_000000_update_player_tables_index.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        });
+
+        Schema::table('player_games', function (Blueprint $table) {
+            $sm = Schema::getConnection()->getDoctrineSchemaManager();
+            $indexesFound = $sm->listTableIndexes('player_games');
+
+            if (!array_key_exists('player_games_game_id_user_id_index', $indexesFound)) {
+                $table->index(['game_id', 'user_id']);
+            }
```
Comment (Jamiras) [inline database/migrations/platform/2023_10_09_000000_update_player_tables_index.php:32]: isn't this handled by the `player_games_user_id_game_id_unique` unique key?

File: database/migrations/platform/2023_10_09_000001_update_achievements_index.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        // add the inverse indexes for user_id so lookups for untracked and deleted are more likely using indexes
+
+        Schema::table('Achievements', function (Blueprint $table) {
+            $sm = Schema::getConnection()->getDoctrineSchemaManager();
+            $indexesFound = $sm->listTableIndexes('Achievements');
+
+            if (!array_key_exists('achievements_game_id_published_index', $indexesFound)) {
+                $table->index(
+                    ['game_id', 'Flags'],
```
Comment (Jamiras) [inline database/migrations/platform/2023_10_09_000001_update_achievements_index.php:20]: This should be `GameID`.
```
SQLSTATE[42000]: Syntax error or access violation: 1072 Key column 'game_id' doesn't exist in table (SQL: alter table `Achievements` add index `achievements_game_id_published_index`(`game_id`, `Flags`))
```

Comment (luchaos) [reply database/migrations/platform/2023_10_09_000001_update_achievements_index.php]: Thanks! Did those blind, should've double checked what I did in the database directly.
Fixed in https://github.com/RetroAchievements/RAWeb/pull/1910/commits/1e1117be4a9bcfa4d4893eca34eeb1785caaaef2

File: database/migrations/platform/2023_10_09_000000_update_player_tables_index.php (hunk 1 above)
Comment (luchaos) [reply database/migrations/platform/2023_10_09_000000_update_player_tables_index.php:32]: That's what I thought! After some tinkering and `EXPLAIN` statements (added them to the PR description above) I found that the query optimizer will only consider the indexes to be used when they exist in this particular order.
Thinking about for a while I guess it makes sense but would've expected it to use the user_id_game_id unique index, too.

**Review by Jamiras — APPROVED** [review]

**Pull request merged**